README.md

```diff
@@ -2,11 +2,110 @@
+
+# Primary Functionality
+
+## Automatic suppression of gradle deprecation warnings
+The nebula-test framework will automatically fail a test if it detects gradle deprecation messages in the output.  Unfortunately, this will fail if another plugin used in a test is using deprecated gradle components even if the actual plugin under test is clean.  The plugin-testing plugin will automatically set the `ignoreDeprecations` system property so that the nebula-test framework will ignore deprecation messages.  Use other mechanisms, such as java deprecation linting, to detect and fix deprecations in the plugin under test.
+
+This behavior can be overridden using the `gradleTestUtils` extension.  e.g.
+
+```groovy
+gradleTestUtils {
+    ignoreGradleDeprecations = false
+}
+```
+
+## Resolution of dependencies in generated build files
+Integration tests for gradle plugins often write build files which include other plugins.  The version of those included plugins is hardcoded in some way, either directly in the test, e.g.
+
+```groovy
+//...within a nebula spec...
+  buildFile << """
+     buildscript {
+         repositories {
+             mavenCentral()
+         }
+         dependencies {
+             classpath 'com.palantir.gradle.conjure:gradle-conjure:0.0.1'
+         }
+     }
+     dependencies {
+         implementation 'com.palantir.conjure.java:conjure-lib:0.0.1'
+     }
+     """
+```
+
+or in some "Versions" class that has a listing of all the dependencies the test uses.  e.g. 
+```java
+final class TestPluginVersions {
+    static final String CONJURE_JAVA = "com.palantir.conjure.java:conjure-java:5.7.1";
+    static final String CONJURE = "com.palantir.conjure:conjure:4.10.1";
+```
+
+Once these tests are written, the versions of the plugins are often not updated, even when the project under test keeps its dependencies of those plugins up-to-date.  This can cause tests to fail when the plugin is updated not because the plugin is bad, but because there is an incompatibility in the old versions of plugins used in the integration tests.  This can often happen with gradle version bumps.
+
+When applied to a project, the plugin-testing plugin scans the `testRuntimeClasspath` configuration for the project and passes all dependencies to the test task as a system property.  The version of the dependencies can then be resolved when tests are run and written into generated files.  e.g.
+
+```groovy
+import static com.palantir.gradle.plugintesting.TestDepVersions.resolve
+import nebula.test.IntegrationSpec
+
+class HelloWorldSpec extends IntegrationSpec {
+   def setup() {
+       //language=gradle
+       buildFile << """
+        buildscript {
+            repositories {
+                mavenCentral()
+            }
+            dependencies {
+                classpath '${resolve('com.palantir.gradle.conjure:gradle-conjure')}'
```

Comment:
The only downside here is (I'm pretty sure) that you get errors when the `//language=gradle` is applied here due to the string interpolation. I wonder if we could just call a method eg:

```groovy
addBuildscriptClasspathDeps(
    'com.palantir.gradle.conjure:gradle-conjure',
    'com.palantir.gradle.jdks:gradle-jdks')
```

that would then add a whole block:

```gradle
buildscript {
    dependencies {
        classpath ...
        classpath ...
    }
}
```

which avoids the red squiggles from interpolation + language insert.

Reply:
Could even do `buildFile << buildscriptClasspathDeps(...)` to keep it flexible for subprojects etc.

Reply:
I figure we'll develop some future fluency calls as we use it more.  You are correct that you get some red squiggles in intellij right now.  I'll look at adding something nicer but would like to get the initial PR through

Reply:
so nebula have an `init.gradle` file that they use as an init script which contain all the dependencies for the buildscript, it's also how you can get your plugin loaded withou having to write a classpath method yourself.

`init.gradle`:
```gradle
allprojects {
    buildscript {
        dependencies {
            classpath files('/Users/fdesouza/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-xml/2.17.2/ad58f5bd089e743ac6e5999b2d1e3cf8515cea9a/jackson-dataformat-xml-2.17.2.jar')
            classpath files('/Users/fdesouza/.gradle/caches/modules-2/files-2.1/com.palantir.human-readable-types/human-readable-types/1.7.0/c72a0a35db41ce785da5669ed520efd8da944456/human-readable-types-1.7.0.jar')
            // ... more dependencies
        }
    }
}
```

What if we extended said `init.gradle` or added our own, where we set constraints on the versions on the buildscript. Then it means we should be able to just say:

```gradle
buildscript {
  dependencies {
    classpath 'com.palantir.gradle.conjure:gradle-conjure'
  }
}
```

> [!NOTE]
> I haven't actually tried this, I'm just putting together building blocks that feel like they should go together.

I have also been thinking about some sort of gcv for the whole init/settings/buildSrc/buildscript space, and it's somewhat tangentially related to this. Just that the `.props` file is from the project, not the project-under-test.

Reply:
I'm also curious with what happens when the main project doesn't actually depend on a dependency, and it's only being included for a particular test. Does this explicitly not cover that case i.e. just write out the version yourself and don't expect it to be up-to-date? If so, it's a bit sad if you do depend on it eventually, you may or may not get the version it seems.

Reply:
in that case the author should add a testRuntimeOnly dependency on the version.  That's what these all are - things the tests depend on at runtime.

Reply:
The init.gradle idea is interesting. I wanted to not be 100% tied to only working in nebula tests since we do have a few repos which use straight gradle toolkit stuff.  But it sounds like a nice enhancement for fluency like Callum's.